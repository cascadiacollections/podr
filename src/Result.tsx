import { h, FunctionComponent } from 'preact';

function formatDuration(duration: number): string {
  return new Date(1000 * duration).toISOString().substr(11, 8);
}

function formatPubDate(isoString: string): string {
  return new Date(isoString.replace(/-/g, "/")).toDateString();
}

function stripHtml(html: string): string {
  const doc: Document = new DOMParser().parseFromString(html, 'text/html');
  return doc.body.textContent || '';
}

interface IEnclosure {
  duration: number;
  link: string;
}

export interface IFeedItem {
  guid: string;
  title: string;
  description: string;
  pubDate: string;
  enclosure: IEnclosure;
}

interface IResultProps {
  result: IFeedItem;
  onClick: (feedItem: IFeedItem) => void;
}

/* tslint:disable:variable-name*/
export const Result: FunctionComponent<IResultProps> = (props: IResultProps) => {
  const { onClick, result } = props;

  return (
    <li
      class={'result list-group-item list-group-item-action'}
      onClick={() => onClick(result)}
      tabIndex={0}>
      <h2 class='title' dangerouslySetInnerHTML={{ __html: result.title }} />
      <strong class='pubDate'>{formatPubDate(result.pubDate)}</strong>
      <strong>&nbsp;&bull;&nbsp;</strong>
      <strong class='duration'>{formatDuration(result.enclosure.duration)}</strong>
      <p class='description' dangerouslySetInnerHTML={{ __html: result.description }} />
    </li>
  );
};
