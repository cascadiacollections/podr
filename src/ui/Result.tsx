import { h, FunctionComponent } from 'preact';

function formatDuration(duration: number): string {
  return new Date(1000 * duration).toISOString().substr(11, 8);
}

function formatPubDate(isoString: string): string {
  return new Date(isoString.replace(/-/g, "/")).toDateString();
}

interface IEnclosure {
  duration: number;
  link: string;
}

export interface IFeedItem {
  readonly guid: string;
  title: string;
  description: string;
  pubDate: string;
  enclosure: IEnclosure;
}

export interface IResultProps {
  result: Readonly<IFeedItem>;
  onClick: (feedItem: Readonly<IFeedItem>) => void;
}

export const Result: FunctionComponent<IResultProps> = (props: IResultProps) => {
  const { onClick, result } = props;
  const { description, title } = result;

  return (
    <li
      class={'result list-group-item list-group-item-action'}
      onClick={() => onClick(result)}
      tabIndex={0}>
      <a href={result.enclosure.link}>
        <h2 class='title' dangerouslySetInnerHTML={{ __html: title }} />
      </a>
      <strong class='pubDate'>{formatPubDate(result.pubDate)}</strong>
      <strong>&nbsp;&bull;&nbsp;</strong>
      <strong class='duration'>{formatDuration(result.enclosure.duration)}</strong>
      <p class='description' dangerouslySetInnerHTML={{ __html: description }} />
    </li>
  );
};
