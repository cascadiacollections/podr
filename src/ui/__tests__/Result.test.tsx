import { h } from 'preact';
import { render } from '@testing-library/preact';
import { Result, IFeedItem } from '../Result';

describe('Result component', () => {
  const mockFeedItem: IFeedItem = {
    guid: 'test-guid-123',
    title: 'Test Episode Title',
    description: 'This is a test episode description',
    pubDate: '2023-05-15T12:00:00Z',
    enclosure: {
      duration: 3600, // 1 hour
      link: 'https://example.com/episode.mp3'
    }
  };

  test('renders correctly', () => {
    const onClick = jest.fn();
    // We need to render the component inside a table/tbody since it's a tr
    const { container } = render(
      <table>
        <tbody>
          <Result result={mockFeedItem} onClick={onClick} />
        </tbody>
      </table>
    );
    expect(container).toMatchSnapshot();
  });

  test('renders with empty duration and date', () => {
    const onClick = jest.fn();
    const emptyFeedItem = {
      ...mockFeedItem,
      pubDate: '',
      enclosure: {
        ...mockFeedItem.enclosure,
        duration: 0
      }
    };
    
    const { container } = render(
      <table>
        <tbody>
          <Result result={emptyFeedItem} onClick={onClick} />
        </tbody>
      </table>
    );
    expect(container).toMatchSnapshot();
  });
});