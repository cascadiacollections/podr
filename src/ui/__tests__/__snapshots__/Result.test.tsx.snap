// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result component renders correctly 1`] = `
<div>
  <table>
    <tbody>
      <tr
        aria-label="Play episode: Test Episode Title"
        role="button"
        tabindex="0"
      >
        <td>
          <a
            aria-label="Stream or download: Test Episode Title"
            href="https://example.com/episode.mp3"
          >
            Test Episode Title
          </a>
        </td>
        <td
          class="date-column"
        >
          <time
            datetime="2023-05-15T12:00:00Z"
          >
            Invalid Date
          </time>
        </td>
        <td
          class="duration-column"
        >
          01:00:00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Result component renders with empty duration and date 1`] = `
<div>
  <table>
    <tbody>
      <tr
        aria-label="Play episode: Test Episode Title"
        role="button"
        tabindex="0"
      >
        <td>
          <a
            aria-label="Stream or download: Test Episode Title"
            href="https://example.com/episode.mp3"
          >
            Test Episode Title
          </a>
        </td>
        <td
          class="date-column"
        >
          <time
            datetime=""
          >
            
          </time>
        </td>
        <td
          class="duration-column"
        >
          00:00:00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
